Koden använder funktionen classification_report från ML-biblioteket scikit-learn för att utvärdera hur bra en (hypotetisk) klassificeringsmodell har presterat. Den jämför de faktiska, sanna värdena (y_true) med de värden som modellen har gissat/predikterat (y_pred).

In [1]:
from sklearn.metrics import classification_report

# De faktiska, sanna klasserna för 5 observationer
y_true = [0, 1, 2, 2, 2]

# De klasser som vår modell gissade på
y_pred = [0, 0, 2, 2, 1]

# Namn på klasserna för att göra utskriften lättare att läsa
target_names = ['class 0', 'class 1', 'class 2']

# Skriv ut utvärderingsrapporten
print(classification_report(y_true, y_pred, target_names=target_names))

              precision    recall  f1-score   support

     class 0       0.50      1.00      0.67         1
     class 1       0.00      0.00      0.00         1
     class 2       1.00      0.67      0.80         3

    accuracy                           0.60         5
   macro avg       0.50      0.56      0.49         5
weighted avg       0.70      0.60      0.61         5



Den utskrivna rapporten ger en detaljerad bild av hur väl modellen presterar, både för varje enskild klass och som en helhet. Till att börja med visar värdet för support hur många faktiska observationer som fanns av varje klass i vårt data. I det här fallet hade vi en observation av klass 0, en av klass 1 och tre observationer av klass 2, vilket ger totalt fem observationer.

För klass 0 lyckades modellen hitta den enda sanna observationen som fanns, vilket ger en recall på 100 procent (1.00). Däremot var precisionen för denna klass endast 50 procent (0.50). Anledningen är att modellen gissade på klass 0 vid två tillfällen, men hade fel den ena gången då det egentligen rörde sig om en etta. När det gäller klass 1 misslyckades modellen totalt. Den gissade att den faktiska ettan var en nolla, och när den väl gissade på en etta var det i själva verket en tvåa. Detta resulterar i både noll i precision och noll i recall för den klassen. För klass 2 var modellen däremot extremt träffsäker när den väl gjorde en förutsägelse; varje gång den gissade på klass 2 hade den rätt, vilket ger en precision på 100 procent (1.00). Dess recall för klass 2 landade dock på cirka 67 procent (0.67), eftersom modellen hittade två av de tre sanna tvåorna, men missade den sista som den istället trodde var en etta. Det redovisade F1-värdet (f1-score) fungerar här som ett sammanvägt betyg av både precision och recall för att ge en mer balanserad bild av prestationerna per klass.

Tittar vi på modellens övergripande prestation ser vi att den gissade rätt på tre av de totalt fem observationerna, vilket ger en total träffsäkerhet (accuracy) på 60 procent (0.60). För att summera prestationerna över alla klasser används sedan två olika medelvärden längst ner i rapporten. Macro-medelvärdet (macro avg) beräknar ett rakt snitt av resultaten för de tre klasserna utan att ta hänsyn till att de är olika stora, vilket till exempel ger en genomsnittlig precision på 50 procent. Det viktade medelvärdet (weighted avg) tar istället hänsyn till antalet observationer (support). Eftersom klass 2 utgör majoriteten av datasetet, väger dess goda resultat mycket tyngre och drar därför upp det sammanlagda viktade snittet för precisionen till 70 procent.